File: src/graphql/mutations/billing/disputes.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// DISPUTE MUTATIONS
// ============================================================================

/**
 * Creates a new billing dispute.
 * @param payload The input data for creating the dispute.
 */
export const createDispute = graphql(`
  mutation CreateDispute($payload: CreateDisputeInput!) {
    billing {
      createDispute(payload: $payload) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the invoice line item ID of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param lineItemId The new invoice line item ID.
 */
export const updateDisputeLineItemId = graphql(`
  mutation UpdateDisputeLineItemId($id: UUID!, $lineItemId: UUID!) {
    billing {
      updateDisputeLineItemId(id: $id, lineItemId: $lineItemId) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the client ID of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param clientId The new client ID.
 */
export const updateDisputeClientId = graphql(`
  mutation UpdateDisputeClientId($id: UUID!, $clientId: UUID!) {
    billing {
      updateDisputeClientId(id: $id, clientId: $clientId) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the reason of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param reason The new reason.
 */
export const updateDisputeReason = graphql(`
  mutation UpdateDisputeReason($id: UUID!, $reason: String!) {
    billing {
      updateDisputeReason(id: $id, reason: $reason) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the status of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param status The new status.
 */
export const updateDisputeStatus = graphql(`
  mutation UpdateDisputeStatus($id: UUID!, $status: DisputeStatusEnum!) {
    billing {
      updateDisputeStatus(id: $id, status: $status) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the disputed amount of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param disputedAmount The new disputed amount.
 */
export const updateDisputeDisputedAmount = graphql(`
  mutation UpdateDisputeDisputedAmount($id: UUID!, $disputedAmount: Float) {
    billing {
      updateDisputeDisputedAmount(id: $id, disputedAmount: $disputedAmount) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the resolution notes of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param resolutionNotes The new resolution notes.
 */
export const updateDisputeResolutionNotes = graphql(`
  mutation UpdateDisputeResolutionNotes($id: UUID!, $resolutionNotes: String) {
    billing {
      updateDisputeResolutionNotes(id: $id, resolutionNotes: $resolutionNotes) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the submitted at timestamp of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param submittedAt The new submitted at timestamp.
 */
export const updateDisputeSubmittedAt = graphql(`
  mutation UpdateDisputeSubmittedAt($id: UUID!, $submittedAt: DateTime) {
    billing {
      updateDisputeSubmittedAt(id: $id, submittedAt: $submittedAt) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the resolved at timestamp of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param resolvedAt The new resolved at timestamp.
 */
export const updateDisputeResolvedAt = graphql(`
  mutation UpdateDisputeResolvedAt($id: UUID!, $resolvedAt: DateTime) {
    billing {
      updateDisputeResolvedAt(id: $id, resolvedAt: $resolvedAt) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the resolved by user ID of a billing dispute.
 * @param id The UUID of the dispute to update.
 * @param resolvedByUserId The new resolved by user ID.
 */
export const updateDisputeResolvedByUserId = graphql(`
  mutation UpdateDisputeResolvedByUserId($id: UUID!, $resolvedByUserId: UUID) {
    billing {
      updateDisputeResolvedByUserId(id: $id, resolvedByUserId: $resolvedByUserId) {
        id
        reason
        status
        disputedAmount
        resolutionNotes
        submittedAt
        resolvedAt
        createdAt
        updatedAt
        invoiceLineItem {
          id
          description
        }
        client {
          id
          name
        }
        resolvedBy {
          id
          name
        }
      }
    }
  }
`);

/**
 * Removes a billing dispute by its ID.
 * @param id The UUID of the dispute to remove.
 */
export const removeDispute = graphql(`
  mutation RemoveDispute($id: UUID!) {
    billing {
      removeDispute(id: $id)
    }
  }
`);

File: services/graphql-billing/src/mutation/disputes.rs
use async_graphql::{Context, InputObject, Object};
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::models::{disputes, enums::DisputeStatusEnum};

#[derive(Debug, Clone, InputObject)]
pub struct CreateDisputeInput {
    pub line_item_id: Uuid,
    pub client_id: Uuid,
    pub reason: String,
    pub status: Option<DisputeStatusEnum>,
    pub disputed_amount: Option<f64>,
    pub resolution_notes: Option<String>,
    pub submitted_at: Option<DateTime<Utc>>,
    pub resolved_at: Option<DateTime<Utc>>,
    pub resolved_by_user_id: Option<Uuid>,
}

#[derive(Debug, Clone, Default)]
pub struct DisputesMutation;

#[Object(name = "BillingDisputesMutation")]
impl DisputesMutation {
    async fn create_dispute(
        &self,
        ctx: &Context<'_>,
        payload: CreateDisputeInput,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "insert into billing.disputes (line_item_id, client_id, reason, status, disputed_amount, resolution_notes, submitted_at, resolved_at, resolved_by_user_id) values ($1, $2, $3, $4, $5, $6, $7, $8, $9) returning *",
        )
        .bind(payload.line_item_id)
        .bind(payload.client_id)
        .bind(payload.reason)
        .bind(payload.status)
        .bind(payload.disputed_amount)
        .bind(payload.resolution_notes)
        .bind(payload.submitted_at)
        .bind(payload.resolved_at)
        .bind(payload.resolved_by_user_id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_line_item_id(
        &self,
        ctx: &Context<'_>,
        line_item_id: Uuid,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set line_item_id = $1 where id = $2 returning *",
        )
        .bind(line_item_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_client_id(
        &self,
        ctx: &Context<'_>,
        client_id: Uuid,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set client_id = $1 where id = $2 returning *",
        )
        .bind(client_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_reason(
        &self,
        ctx: &Context<'_>,
        reason: String,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set reason = $1 where id = $2 returning *",
        )
        .bind(reason)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_status(
        &self,
        ctx: &Context<'_>,
        status: DisputeStatusEnum,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set status = $1 where id = $2 returning *",
        )
        .bind(status)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_disputed_amount(
        &self,
        ctx: &Context<'_>,
        disputed_amount: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set disputed_amount = $1 where id = $2 returning *",
        )
        .bind(disputed_amount)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_resolution_notes(
        &self,
        ctx: &Context<'_>,
        resolution_notes: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set resolution_notes = $1 where id = $2 returning *",
        )
        .bind(resolution_notes)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_submitted_at(
        &self,
        ctx: &Context<'_>,
        submitted_at: Option<DateTime<Utc>>,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set submitted_at = $1 where id = $2 returning *",
        )
        .bind(submitted_at)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_resolved_at(
        &self,
        ctx: &Context<'_>,
        resolved_at: Option<DateTime<Utc>>,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set resolved_at = $1 where id = $2 returning *",
        )
        .bind(resolved_at)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_dispute_resolved_by_user_id(
        &self,
        ctx: &Context<'_>,
        resolved_by_user_id: Option<Uuid>,
        id: Uuid,
    ) -> async_graphql::Result<disputes::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, disputes::Model>(
            "update billing.disputes set resolved_by_user_id = $1 where id = $2 returning *",
        )
        .bind(resolved_by_user_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn remove_dispute(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query("delete from billing.disputes where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new("Unable to remove dispute"));
        }

        Ok("Dispute removed successfully".into())
    }
}
